ADD DDPG Example for Pendulum Problem

amifunny opened the pull request:
Added DDPG example with Target Networks and Experience Buffer on Pendulum Problem. All lines are within 90 char limit and everything looks good. But i was not able to run this command to see generated file `python autogen.py add_example rl/DDPG_Pendulum`. But everything else in .py is good.
Please review.

fchollet commented on examples/rl/DDPG_Pendulum.py:
```diff
@@ -0,0 +1,428 @@
+actor_optimizer = tf.keras.optimizers.Adam(a_lr)
+
+total_eps = 100
+GAMMA = 0.99
```
Why is this variable in all caps and the other lowercase? Use all caps for global constants defined at the start of the file only.

fchollet commented on examples/rl/DDPG_Pendulum.py:
```diff
@@ -0,0 +1,428 @@
+
+def get_critic():
+
+    init = tf.keras.initializers.GlorotNormal(seed=1)
```
You can remove the seed; the code is not deterministic in its current state. Just use the string "glorot_normal".

Question: does this initializer really help, compared to the default (uniform)?

amifunny commented:
Thanks for the Review. I have made all the changes that you pointed out. Plus some additional changes that i thought were better-

- Added new graph image with tau=0.005 as this show more incremental learning than the previous vertical graph.
- Removed seed initializers and allowed default 'GlorotUniform', and it did not affected learning.
- Added more comments about the Actor-Critic Losses
- Replaced LayerNorm with BatchNorm which i found on again reviewing the paper.
- Also new full trained gif , that shows that agent learned to make pendulum upright.
- And made the target parameter update function more concise.

Additional Comments - 

- It was silly of me to use personalized comments. I have Removed and Corrected ALL. But i hope its ok to use "We" in comments.
- changed mu and sigma to mean and std_dev but other dt and x0 are kept as it is as they are not very important variables and go perfectly with the formula.
- Also suggestion at last are not changed which i thought will benefit the reader.
- Variable name like init and grad are not changed as they easily convey the meaning.

amifunny commented:
@fchollet If there are more changes that you would like to have , please let me know.

fchollet commented on examples/rl/DDPG_Pendulum.py:
```diff
@@ -0,0 +1,414 @@
+        self.batch_size = batch_size
+
+        # Its tells us num of times record() was called.
+        self.buffer_ctr = 0
```
Again, prefer using fully-spelled out variable names such as "counter", to improve readability

amifunny commented:
@fchollet Thanks for your Input. I have more polished the example to the extent of my grammatical and English knowledge. And code is as concise as it gets. And Looked at other keras examples for inspiration.

Still if code lacks something, let me know.

amifunny commented:
Thanks for Review and precious Inputs. It would be a great deal for me to have my example accepted. 

I tried `python autogen.py add_example vision/script_name` on windows system which showed error of `No such file or directory: '/tmp/5818881.py` which I guess is due to file permission settings on my system.
**So I used autogen command on colab and only copied  - .md , .ipynb and img folder into master. So if there are other autogenerated files these are not copied in this commit.**

And I was also not able to preview my example using `python autogen.py make` as error `UnicodeDecodeError: 'charmap' codec can't decode byte` was prompted.

**But I noticed that autogen script captured the new graph image, which will conflict with online image of graph already in this example and also output of training 100 episodes was captured in `.md` file which might not be good to see 100 lines of log in an example.**

I did not tampered with any autogenerated files just the thing I noticed!

amifunny commented:
All changes done. And please let me know, which graph will you go with.
